File: iosApp/iosApp/Presentation/Game/GameView.swift
//
//  GameView.swift
//  iosApp
//
//  Created by ウルトラ深瀬 on 2026/05/31.
//

import SwiftUI
import SharedLogic

struct GameView: View {
    @State var viewModel: GameViewModel = GameViewModel(weaponResourceGetUseCase: WeaponResourceGetUseCase(weaponRepository: WeaponRepository(weaponDataSource: WeaponDataSource())))
    
    var body: some View {
        VStack(spacing: 0) {
            Spacer().frame(height: 12)
            
            HStack {
                Text("Loaded : [ \(loadedWeaponsText) ]")
                    .font(.system(size: 22, weight: .bold))
                
                Spacer()
            }
            
            Spacer().frame(height: 16)
            
            weaponDisplayArea(viewModel.currentWeapon)
            
            Spacer()
            
            actionButtons
            
            Spacer().frame(height: 32)
        }
        .padding(.horizontal, 32)
        .background(.black)
        .foregroundStyle(.green)
    }
    
    var loadedWeaponsText: String {
        return viewModel.loadedWeapons.map(\.name).joined(separator: ", ")
    }
    
    @ViewBuilder
    func weaponImage(name: String?) -> some View {
        Group {
            if let name = name {
                Image(name)
                    .resizable()
                    .renderingMode(.template)
                    .padding(24)
            } else {
                if viewModel.isLoading {
                    ProgressView()
                        .progressViewStyle(.circular)
                        .tint(.white)
                        .scaleEffect(2)
                    
                } else {
                    Text("Select any weapon.")
                        .font(.system(size: 28, weight: .bold))
                }
            }
        }
        .frame(maxWidth: .infinity, maxHeight: .infinity)
        .background(.green)
        .foregroundStyle(.white)
        .aspectRatio(1, contentMode: .fit)
    }
    
    func specText(title: String, value: String) -> some View {
        HStack {
            Text(title)
                .font(.system(size: 28, weight: .bold))
            
            Spacer()
            
            Text(value)
                .font(.system(size: 28, weight: .bold))
        }
    }
    
    func weaponDisplayArea(_ weapon: Weapon?) -> some View {
        let capacityText: String = {
            if let capacity = weapon?.capacity {
                return "\(capacity) bullets"
            } else {
                return " - "
            }
        }()
        return VStack(alignment: .leading) {
            weaponImage(name: weapon?.imageName)
                .aspectRatio(1, contentMode: .fit)
            
            specText(title: "Name : ", value: weapon?.name ?? " - ")
            specText(title: "Capacity : ", value: capacityText)
        }
    }
    
    func actionButton(title: String, action: @escaping () -> Void) -> some View {
        Button {
            action()
        } label: {
            Text(title)
                .font(.system(size: 24, weight: .bold))
                .padding()
                .foregroundStyle(.black)
                .frame(maxWidth: .infinity)
                .frame(height: 52)
                .background(.white)
                .clipShape(RoundedRectangle(cornerRadius: 10))
        }
    }
    
    var actionButtons: some View {
        VStack(spacing: 0){
            HStack(spacing: 0){
                actionButton(title: "Pistol") {
                    viewModel.weaponSelected(id: 0)
                }
                
                Spacer().frame(width: 12)
                
                actionButton(title: "Bazooka") {
                    viewModel.weaponSelected(id: 1)
                }
            }
            
            Spacer().frame(height: 12)
                      
            actionButton(title: "Reset") {
                viewModel.resetButtonTapped()
            }
        }
    }
}

#Preview {
    GameView()
}

File: androidApp/src/main/kotlin/org/example/project/presentation/GameView.kt
package org.example.project.presentation

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.aspectRatio
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.safeContentPadding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Icon
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import org.example.project.domain.entities.Weapon
import androidx.compose.ui.platform.LocalResources

@Composable
fun GameView(
    viewModel: GameViewModel
) {
    val state by viewModel.uiState.collectAsStateWithLifecycle()
    val loadedWeaponsText: String = state.loadedWeapons.joinToString(separator = ", ") { it.name }

    Surface(
        color = Color.Black,
        modifier = Modifier
            .fillMaxSize(),
    ) {
        Column(
            modifier = Modifier
                .safeContentPadding()
        ) {
            Spacer(modifier = Modifier.height(12.dp))

            Row {
                Text(
                    "Loaded : [ $loadedWeaponsText ]",
                    fontSize = 22.sp,
                    fontWeight = FontWeight.Bold,
                    color = Color.Green,
                )

                Spacer(modifier = Modifier.weight(1f))
            }

            Spacer(modifier = Modifier.height(16.dp))

            WeaponDisplayArea(
                weapon = state.currentWeapon,
                isLoading = state.isLoading,
            )

            Spacer(modifier = Modifier.weight(1f))

            ActionButtons(viewModel = viewModel)

            Spacer(modifier = Modifier.height(32.dp))
        }
    }
}

@Composable
fun WeaponImage(name: String?, isLoading: Boolean) {
    Box(
        modifier = Modifier
            .background(Color.Green)
            .aspectRatio(
                ratio = 1f,
                matchHeightConstraintsFirst = true
            ),
        contentAlignment = Alignment.Center,
    ) {
        if (name != null) {
            Icon(
                painter = painterResource(
                    id = LocalResources.current.getIdentifier(
                        name,
                        "drawable",
                        LocalContext.current.packageName
                    )
                ),
                contentDescription = "Weapon Icon",
                tint = Color.White,
                modifier = Modifier.padding(24.dp)
            )

        } else {
            if (isLoading) {
                CircularProgressIndicator(color = Color.White)

            } else {
                Text(
                    "Select any weapon.",
                    fontSize = 28.sp,
                    fontWeight = FontWeight.Bold,
                    color = Color.White
                )
            }
        }
    }
}

@Composable
fun SpecText(title: String, value: String) {
    Row {
        Text(
            title,
            fontSize = 28.sp,
            fontWeight = FontWeight.Bold,
            color = Color.Green,
        )

        Spacer(modifier = Modifier.weight(1f))

        Text(
            value,
            fontSize = 28.sp,
            fontWeight = FontWeight.Bold,
            color = Color.Green,
        )
    }
}

@Composable
fun WeaponDisplayArea(weapon: Weapon?, isLoading: Boolean) {
    val capacityText = weapon?.capacity?.let { "$it bullets" } ?: " - "
    Column(
        horizontalAlignment = Alignment.Start,
    ) {
        WeaponImage(name = weapon?.imageName, isLoading = isLoading)

        Spacer(modifier = Modifier.height(12.dp))

        SpecText(title = "Name : ", value = weapon?.name ?: " - ")

        Spacer(modifier = Modifier.height(4.dp))

        SpecText(title = "Capacity : ", value = capacityText)
    }
}

@Composable
fun ActionButton(
    title: String,
    modifier: Modifier,
    action: (() -> Unit)
) {
    Button(
        onClick = action,
        modifier = modifier
            .height(52.dp),
        colors = ButtonDefaults.buttonColors(
            containerColor = Color.White,
            contentColor = Color.Black,
        ),
        shape = RoundedCornerShape(size = 10.dp)
    ) {
        Text(
            title,
            fontSize = 24.sp,
            fontWeight = FontWeight.Bold,
            color = Color.Black,
        )
    }
}

@Composable
fun ActionButtons(viewModel: GameViewModel) {
    Column(
        modifier = Modifier.fillMaxWidth()
    ) {
        Row {
            ActionButton(
                title = "Pistol",
                modifier = Modifier.weight(1f),
            ) {
                viewModel.weaponSelected(id = 0)
            }

            Spacer(modifier = Modifier.width(12.dp))

            ActionButton(
                title = "Bazooka",
                modifier = Modifier.weight(1f),
            ) {
                viewModel.weaponSelected(id = 1)
            }
        }

        Spacer(modifier = Modifier.height(12.dp))

        ActionButton(
            title = "Reset",
            modifier = Modifier.fillMaxWidth(),
        ) {
            viewModel.resetButtonTapped()
        }
    }
}